src/integration/integration.service.ts

```diff
@@ -0,0 +1,442 @@
+              .includes(coupon.code),
+        );
+      if (applicableCoupons.length === 0) {
+        cartUpdateActions.setInapplicableCoupons(inapplicableCoupons);
+        return;
+      }
+      //We need to do another call to V% if there is any applicable coupon in the cart
+      //to get definitions of discounts we should apply on the cart
+      validatedCoupons =
+        await this.voucherifyConnectorService.validateStackableVouchers(
+          buildValidationsValidateStackableParamsForVoucherify(
+            applicableCoupons,
+            cart,
+            mapItemsToVoucherifyOrdersItems(items),
+          ),
+        );
+    }
+
+    let productsToAdd: ProductToAdd[] = [];
+    if (typeof cartUpdateActions.getProductsToAdd === 'function') {
+      productsToAdd = await cartUpdateActions.getProductsToAdd(
+        validatedCoupons.redeemables.filter(
+          (redeemable) =>
+            redeemable.result?.discount?.type === 'UNIT' &&
+            redeemable.result.discount.unit_type !== FREE_SHIPPING_UNIT_TYPE,
+        ),
+      );
+    }
+
+    const productsToAddWithIncorrectPrice = productsToAdd.filter(
+      (product) => product.discount_difference,
+    );
+
+    if (productsToAddWithIncorrectPrice.length) {
+      const itemsWithPricesCorrected = await this.getItemsWithCorrectedPrices(
+        validatedCoupons.order.items,
+        productsToAddWithIncorrectPrice,
+      );
+      validatedCoupons =
+        await this.voucherifyConnectorService.validateStackableVouchers(
+          buildValidationsValidateStackableParamsForVoucherify(
+            couponsLimited.filter((coupon) => coupon.status != 'DELETED'),
+            cart,
+            itemsWithPricesCorrected,
+          ),
+        );
+    }
+
+    let redeemables = validatedCoupons?.redeemables;
+    if (promotions.length) {
+      redeemables = this.voucherifyService.setBannerOnValidatedPromotions(
+        validatedCoupons,
+        promotions,
+      );
+    }
+
+    this.logger.debug({
+      msg: 'Validated coupons',
+      validatedCoupons,
+      availablePromotions,
+      productsToAdd,
+    });
+
+    if (
+      typeof cartUpdateActions.setSessionKey === 'function' &&
+      typeof cartUpdateActions.setTotalDiscountAmount === 'function' &&
+      typeof cartUpdateActions.setApplicableCoupons === 'function' &&
+      typeof cartUpdateActions.setInapplicableCoupons === 'function' &&
+      typeof cartUpdateActions.setProductsToAdd === 'function'
+    ) {
+      cartUpdateActions.setSessionKey(validatedCoupons?.session?.key);
+      cartUpdateActions.setTotalDiscountAmount(
+        calculateTotalDiscountAmount(validatedCoupons),
+      );
+      cartUpdateActions.setApplicableCoupons(
+        getCouponsByStatus(redeemables, 'APPLICABLE'),
+      );
+      cartUpdateActions.setInapplicableCoupons(inapplicableCoupons);
+      cartUpdateActions.setProductsToAdd(productsToAdd);
+    }
+    return;
+  }
+
+  private async getItemsWithCorrectedPrices(
```

Comment:
it looks more like mapping/merging objects than high level operation. Maybe we will hide those details in some util class?

Reply:
Also... looks like this function has two responsibilities:
- mapping OrdersItem to some unknown structure
- correcting the price

Reply:
- mapping OrdersItem to OrdersItem
- correcting the price
nothing crazy